Update Metadata Documentation

kuzdogan opened the pull request:
- Update the example metadata JSON to follow the alphabetically ordered canonical format. Also add the `"errors"` field in `devdoc` and `userdoc`.

- Shorten the CBOR encoding part and add a link to https://playground.sourcify.dev
  - I think this is demonstrated better in action.
  - Also the previous docs were both giving the example of `0xa264` as well as mentioning multiple times not to use this pattern. It's better to not mention this pattern at all and to explain the correct way to decode.
- Update the explanation for the Source Code Verification section

kuzdogan commented on docs/metadata.rst:
```diff
@@ -153,34 +95,104 @@ explanatory purposes.
+        "compilationTarget": {
+          "myDirectory/myFile.sol": "MyContract"
+        },
+        // Required for Solidity.
+        "evmVersion": "london"
```
Not sure if this is Required but I believe so. Where can I find the schema of the metadata or a validation for that?

nikola-matic replied:
I have no idea, but the best I can do is the [parser/validator used in our tests](https://github.com/ethereum/solidity/blob/develop/test/Metadata.cpp#L194), which would lead me to believe that `evmVersion` is in fact a mandatory field.

nikola-matic commented on docs/metadata.rst:
```diff
@@ -30,108 +30,50 @@ shall match with the one contained in the bytecode.
-reduce whitespace to a minimum and sort the keys of all objects to arrive at a
-unique formatting. Comments are not permitted and used here only for
+reduce whitespace to a minimum and sort the keys of all objects in alphabetical ordering
+to arrive at a canonical formatting. Comments are not permitted and used here only for
```
```suggestion
reduce whitespace to a minimum, and sort the keys of all objects in alphabetical order
to arrive at a canonical formatting. Comments are not permitted and are used here only for
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -200,23 +212,32 @@ explanatory purposes.
+The compiler currently by default appends the
+`IPFS hash (in CID v0) <https://docs.ipfs.tech/concepts/content-addressing/#version-0-v0>`_
+of the canonical metadata file and the compiler version to the end of the bytecode.
+Optionally a Swarm hash instead of the IPFS, or an experimental flag is used.
```
```suggestion
Optionally, a Swarm hash instead of the IPFS, or an experimental flag is used.
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -200,23 +212,32 @@ explanatory purposes.
+`IPFS hash (in CID v0) <https://docs.ipfs.tech/concepts/content-addressing/#version-0-v0>`_
+of the canonical metadata file and the compiler version to the end of the bytecode.
+Optionally a Swarm hash instead of the IPFS, or an experimental flag is used.
+Below are the all possible fields:
```
```suggestion
Below are all the possible fields:
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -252,24 +265,27 @@ is JSON-decoded into a structure like above.
-(`npm package <https://www.npmjs.com/package/source-verify>`_) you can see
-example code that shows how to use this feature.
+`Sourcify <https://sourcify.dev>`_ makes use of this feature for "full/perfect verification",
+as well as pins the files publicly on IPFS to be accessed with the metadata hash.
```
```suggestion
as well as pinning the files publicly on IPFS to be accessed with the metadata hash.
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -30,108 +30,50 @@ shall match with the one contained in the bytecode.
         // Required: ABI definition of the contract. See "Contract ABI Specification"
         "abi": [/* ... */],
-        // Required: NatSpec developer documentation of the contract.
+        // Required: NatSpec developer documentation of the contract. See "NatSpec Format"
```
```suggestion
        // Required: NatSpec developer documentation of the contract. See https://docs.soliditylang.org/en/latest/natspec-format.html for details.
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -153,34 +95,104 @@ explanatory purposes.
+            "deduplicate": false,
+            "cse": false,
+            "constantOptimizer": false,
+            "yul": true,
```
These should be in alphabetical order too, no?

kuzdogan replied:
Yes 👍

nikola-matic commented on docs/metadata.rst:
```diff
@@ -153,34 +95,104 @@ explanatory purposes.
+          "myDirectory/myFile.sol": "MyContract"
+        },
+        // Required for Solidity.
+        "evmVersion": "london"
```
```suggestion
        "evmVersion": "london",
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -153,34 +95,104 @@ explanatory purposes.
+        // Required for Solidity: Addresses for libraries used.
+        "libraries": {
+          "MyLib": "0x123123..."
+        }
```
```suggestion
        },
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -30,108 +30,50 @@ shall match with the one contained in the bytecode.
+            "Transfer(address,address,uint256)": {
+              "details": "Emitted when `value` tokens are moved from one account (`from`) toanother (`to`)."
+              "params": {
+                "from": "The sender address"
+                "to": "The receiver address"
```
```suggestion
                "from": "The sender address",
                "to": "The receiver address",
```

nikola-matic commented on docs/metadata.rst:
```diff
@@ -30,108 +30,50 @@ shall match with the one contained in the bytecode.
+          },
+          "events": {
+            "Transfer(address,address,uint256)": {
+              "details": "Emitted when `value` tokens are moved from one account (`from`) toanother (`to`)."
```
```suggestion
              "details": "Emitted when `value` tokens are moved from one account (`from`) toanother (`to`).",
```

nikola-matic reviewed (commented):
The JSON is malformed due to missing commas (`,`) in certain places - I think I caught all of them, but I'd still suggest running the entire object through something like `jq` just to make sure.
Otherwise, looks good. Fix and squash, and it should be good to merge.

nikola-matic reviewed (dismissed):
Looks good. Please squash the commits before merging.

kuzdogan commented:
Added "appends to **runtime** bytecode" explicitly as it is not necessarily appended to the creation bytecode.

I find it useful to break down the contents of the metadata when explaining into two:
- How to interact with the contract: ABI, and NatSpec documentation.
- How to reproduce the compilation and verify a deployed contract:
  compiler version, compiler settings, and source files used.

Changed the first paragraph to this to quickly give an overview of the **purpose** of the file as well as its **contents**

nikola-matic reviewed (dismissed):
Squash the commits please.

kuzdogan commented:
> Squash the commits please.

Sorry I thought this was meant for the person merging to click "Squash and merge" in the PR. Do I need to squash, i.e. rebase and force-push everything into one commit?

nikola-matic commented:
@kuzdogan yeah, it seems like it's a configurable feature, and I have no idea how we have it configured, so please rebase, squash into one commit, and force push, and I'll merge it.

kuzdogan committed:
Update metadata docs

Squashed into single commit

Squashed commits:
Sort metadata JSON, add errors to devdoc & userdoc

Sorts the example metadata in the alphabetical order,
as this is the way output by the compiler.

Add playground.sourcify.dev link

Update metadata documentation content

Fix trailing whitespaces

Add new line after code-block

Fix unexpected indentation

Fix unexpected unindent

Fix unexpected unindent - 2

Suggestions from code review: wording, punctuation

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>
Fix missing trailing commas

Co-authored-by: Nikola Matić <nikola.matic@ethereum.org>
Order yul settings, fix trailing commas

Explicitly state appended to the runtime bytecode

Break down metadata content into two points

Remove trailing whitespace
